### PR TITLE
Fix timestamp crash and signature in Windows

### PR DESCRIPTION
**Pull request by javierluraschi**
Two fixes for `timestamp()`:

 1. Avoid crashing while using the namespace `utils::timestamp`
 2. Add missing support for `stamp`, `prefix`, `suffix` and `quiet` parameters.

Somewhat related to: https://github.com/rstudio/rstudio/pull/1029

### CODE REVIEW

File: src/cpp/session/modules/SessionHistory.cpp [hunk 1]
```diff
@@ -361,12 +361,20 @@ void onHistoryAdd(const std::string& command)
...
 }
 
-SEXP rs_timestamp(SEXP dateSEXP)
+SEXP rs_timestamp(SEXP stampSEXP, SEXP prefixSEXP, SEXP suffixSEXP, SEXP quietSEXP)
 {
-   boost::format fmt("##------ %1% ------##");
-   std::string ts = boost::str(fmt % r::sexp::safeAsString(dateSEXP));
+   std::string prefix = r::sexp::safeAsString(prefixSEXP);
+   
+   std::string suffix = r::sexp::safeAsString(prefixSEXP);
```
Comment (jmcphers) [inline src/cpp/session/modules/SessionHistory.cpp]: I think this should be `suffixSEXP` :-)

File: src/cpp/session/modules/SessionHistory.cpp [hunk 2]
```diff
@@ -361,12 +361,20 @@ void onHistoryAdd(const std::string& command)
    module_context::enqueClientEvent(event);
 }
 
-SEXP rs_timestamp(SEXP dateSEXP)
+SEXP rs_timestamp(SEXP stampSEXP, SEXP prefixSEXP, SEXP suffixSEXP, SEXP quietSEXP)
 {
-   boost::format fmt("##------ %1% ------##");
-   std::string ts = boost::str(fmt % r::sexp::safeAsString(dateSEXP));
+   std::string prefix = r::sexp::safeAsString(prefixSEXP);
```
Comment (jmcphers) [inline src/cpp/session/modules/SessionHistory.cpp]: There is a risk here that the prefix or suffix will contain characters that will be interpreted by `boost::fmt` and cause the formatter to fail. We could escape them but I think we should just use regular string append/prepend here.

Comment (jjallaire) [inline src/cpp/session/modules/SessionHistory.cpp]: I agree, as I recall an invalid `boost::format` will throw an exception which is basically a system crash in our environment.

File: src/cpp/session/modules/SessionHistory.cpp (hunk 1 above)
Comment (javierluraschi) [inline src/cpp/session/modules/SessionHistory.cpp]: :astonished: Thanks!

File: src/cpp/session/modules/SessionHistory.cpp (hunk 2 above)
Comment (javierluraschi) [inline src/cpp/session/modules/SessionHistory.cpp]: Makes sense, fixed

**Comment by kevinushey** [comment]
Note that this implementation won't handle vectorization properly -- in theory someone might call `timestamp` as e.g.

```
utils::timestamp(c("1", "2"), "3", c("4", "5"))
```

And that works fine with plain R, but will fail here.

I would suggest avoiding a native routine call and just replacing the definition with something like:

```R
function (stamp = date(), prefix = "##------ ", suffix = " ------##",
    quiet = FALSE)
{
    stamp <- paste0(prefix, stamp, suffix)
    if (!quiet)
        cat(stamp, sep = "\n")
    invisible(stamp)
}
```

Note that this is the same definition of `utils::timestamp()`, with the `.External2(C_addhistory, stamp)` call removed (presumedly this is the one causing the crash).

**Comment by javierluraschi** [comment]
That makes a lot of sense Kevin! Not sure if there was a good reason to implement this in native code originally so moving this to R works for me.

**Comment by jmcphers** [comment]
The external call `.External2(C_addhistory, stamp)` is equivalent to what we do here:

https://github.com/rstudio/rstudio/blob/d13a0b9d4cba8e44b7beee0b8f74c8b611bbe3cf/src/cpp/session/modules/SessionHistory.cpp#L368

Timestamps with `quiet=TRUE` can be used to help organize console history by inserting a timestamp into the history (without actually cluttering the console output with one). It's fine to move this back into R but we will still need to call back into RStudio to inject the history entry.

**Pull request merged**